### PR TITLE
Groups

### PR DESCRIPTION
**Pull request by dev-matan-tsuberi**
Groups allow us to assign keys to multiple users at the same time. We can easily build complex groups by using simple operations: fixed group, union, intersection and inverse, as well as create new group types by implementing `isMember(..)` (e.g. all accounts which have at least X eth).
Groups act as a proxy restricted only to members.

Still need to figure out how best to forward the message call.

### CODE REVIEW

**Comment by dev-matan-tsuberi** [comment]
@leviadam @ben-kaufman Does anyone know what's the best way to do forwarding?

**Comment by leviadam** [comment]
@dev-matan-tsuberi, I'm not sure we want to merge this should be part of this repo.

**Comment by dev-matan-tsuberi** [comment]
@leviadam Why not?

**Comment by leviadam** [comment]
@dev-matan-tsuberi Do you feel it should be part of our EIP?
Maybe a second EIP?

**Comment by dev-matan-tsuberi** [comment]
This is not a part of the current EIP, could be a another EIP (not right now though). Its a complementary feature.

**Comment by leviadam** [comment]
OK. We can have both in the same repo and explain in the readme the are complementary.

**Comment by leviadam** [comment]
I think tests and mocks are a good start

File: contracts/Groups.sol [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+    Group group;
+
+    constructor(Group _group) public {
+        group = _group;
+    }
+
+    function isMember(address _account) public view returns(bool) {
+        return !group.isMember(_account);
+    }
+}
```
Comment (orenyodfat) [inline contracts/Groups.sol]: remove CR

File: contracts/Groups.sol [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+    function forward(address _to, bytes4 _selector, bytes _args) public payable onlyMember {
+        /// TODO: Figure out how to best forward the call
+    }
+}
+
+
+contract FixedGroup is Group {
+    mapping(address => bool) members;
+
+    constructor(address[] _members) public {
```
Comment (orenyodfat) [inline contracts/Groups.sol]: Please comment the limitation of _members length

Comment (dev-matan-tsuberi) [reply contracts/Groups.sol]: Anyway to calculate the limitation?

File: contracts/Groups.sol (hunk 1 above)
Comment (dev-matan-tsuberi) [reply contracts/Groups.sol]: https://github.com/daostack/access_control/pull/38/commits/21b6758f3e9476b3e5c4405a2651c37a7e92533f

**Comment by dev-matan-tsuberi** [comment]
https://github.com/daostack/access_control/pull/38/commits/e0718f528a46061b23336dfd6fd171d0b647418f: @leviadam Added forwarding ability, a mock and some tests. @orenyodfat Added function comments.

File: contracts/Groups.sol [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+/**
+ * @title Abstract base contract for all groups.
+ */
+contract Group {
+
+    /**
+     * @dev Check if an account belongs to this group.
+     * @param _account the account to check.
+     */
+    function isMember(address _account) public view returns(bool);
```
Comment (orenyodfat) [inline contracts/Groups.sol]: is this should not be defined as interface ?

File: contracts/Groups.sol [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+
+    /**
+     * @dev Modifier for restricting access to members of the group.
+     */
+    modifier onlyMember() {
+        require(isMember(msg.sender), "Not a member of this group");
+
```
Comment (orenyodfat) [inline contracts/Groups.sol]: redundant line space

File: contracts/Groups.sol [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+}
+
+
+/**
+ * @title Group with fixed members set at creation time.
+ */
+contract FixedGroup is Group {
+    mapping(address => bool) members;
+
+    constructor(address[] _members) public {
```
Comment (orenyodfat) [inline contracts/Groups.sol]: what is the limitation for the number of _members which can be sent to this constructor ?

**Comment by dev-matan-tsuberi** [comment]
https://github.com/daostack/access_control/pull/38/commits/0b20aca662d7b146d28d1d890efad2b2dee00042 @orenyodfat Done

**Pull request merged**